tests/lib/core/t-sequence.c

```diff
@@ -198,6 +198,110 @@ START_TEST (test_bt_sequence_length1)
 }
 END_TEST
 
+/*
+ * test_bt_sequence_length_reduce:
+ *
+ * A sequence has both a "nominal" length and a "pattern" length. The former
+ * dictates when the song stops playing, and the latter represents the index at
+ * which the final non-null piece of sequence data is found. This case asserts
+ * some expected behaviour around setting the nominal song length to be less
+ * than the pattern length, i.e. when the user uses "CTRL-E" to set the song end
+ * to be halfway through an existing song. If the implementation is incorrect,
+ * then sequence data will be truncated.
+ */
+START_TEST (test_bt_sequence_length_reduce)
+{
+  BT_TEST_START;
+  
+  BtSequence *sequence =
+      (BtSequence *) check_gobject_get_object_property (song, "sequence");
+  g_object_set (sequence, "length", 8L, NULL);
+  
+  BtMachineConstructorParams cparams;
+  cparams.id = "genm";
+  cparams.song = song;
+  
+  BtMachine *machine = BT_MACHINE (bt_source_machine_new (&cparams,
+          "buzztrax-test-mono-source", 0L, NULL));
+
+  BtCmdPattern *pattern =
+      (BtCmdPattern *) bt_pattern_new (song, "pattern-name", 8L, machine);
+  
+  bt_sequence_add_track (sequence, machine, -1);
+ 
+  for (gint i = 0; i < 8; ++i) {
+    bt_sequence_set_pattern (sequence, i, 0, pattern);
+  }
+  
+  /* Set the nominal length to be half the pattern length, and check that length
+     values are correctly updated and that no sequence data has been lost. */
+  GST_INFO ("-- act --");
+  g_object_set (sequence, "length", 4L, NULL);
+  g_object_set (sequence, "len-patterns", 8L, NULL);
+
+  GST_INFO ("-- assert --");
```

Comment:
Let me know if I am too picky here, but imho this test is too long and checks too many things at ones. This is not great because:
* once the first assert fails, the assert below are not run
* the name of the failing test does not give you a good clue anymore

Can you try to split the test? Eg. testing that the length is 4 after setting a length to 4 and then a `len-pattern` to 8 could be a test of its own. If you need similar code for setup, this can be extracted to a local helper.

Ideally a test has only a single arrange/act/assert sequence.

Reply:
Thanks for the reference. I'll read it over the weekend. The book I read many years ago that led me to rewrite most of the tests was: https://www.artofunittesting.com/.

I don't necessarily mind larger tests, but then ideally prerequisites are covered by unit tests and the larger test only runs, when the prerequisites have been verified.

Reply:
I really like this response https://henrikwarne.com/2014/09/04/a-response-to-why-most-unit-testing-is-waste/. E.g. when writing a function that generate unique machine ids - that can be perfectly unit-tested. I find integration test fragile and error prone if low level stuff is not caught by unit tests before hand. Anyway the PR review is not a good forum for discussio - I should probably enable https://github.com/Buzztrax/buzztrax/discussions/landing or we find a different place to continue this.